nurse_scheduling.py

```diff
@@ -0,0 +1,162 @@
+min_duty_days = int(n_days/n_nurses)
+
+
+# Find index into 1D list for (nurse_index, day_index)
+def get_index(nurse_index, day_index):
+    return nurse_index * n_days + day_index
+
+
+# Inverse of get_index - given an index in a 1D list, return the nurse_index
+# and day_index
+def get_nurse_and_day(index):
+    nurse_index, day_index = divmod(index, n_days)
+    return nurse_index, day_index
+
+
+# Hard nurse constraint: no nurse works two consecutive days
+# It does not have Lagrange parameter - instead, J matrix
+# symmetric, real-valued interaction matrix J, whereas all terms are
+# a or zero.
+# composite indices i(n, d) and j(n, d) as functions of n and d
+# In other words, the indices are not n and d. They are kind of
+# (n+d)**2, and we need to take that into account. This seems to be
+# a clean way to do it.
+# J_i(n,d)j(n,d+1) = a and 0 otherwise.
+J = defaultdict(int)
+for nurse in range(n_nurses):
+    for day in range(n_days - 1):
+        nurse_day_1 = get_index(nurse, day)
+        nurse_day_2 = get_index(nurse, day+1)
+        J[nurse_day_1, nurse_day_2] = a
+
+# Q matrix assign the cost term, the J matrix
+Q = deepcopy(J)
+
+# Hard shift constraint. The sum is over each day.
+# lagrange_parameter * ((sum(effort * q) - workforce) ** 2)
+for nurse_day_1 in range(size):
+    _, date_index = get_nurse_and_day(nurse_day_1)
+    # Diagonal term, without the workforce * workforce
+    Q[nurse_day_1, nurse_day_1] += lagrange_parameter * (1 - (2 * workforce))
+    for nurse_day_2 in range(size):
+        _, day_index_2 = get_nurse_and_day(nurse_day_2)
+        # Include only the same day, across nurses
+        if (date_index == day_index_2 and nurse_day_2 != nurse_day_1):
+            Q[nurse_day_1, nurse_day_2] += lagrange_parameter * 2
+
+# Soft nurse constraint
+# gamma * ((sum(h * q) - min_duty_days) ** 2)
+for nurse_day_1 in range(size):
+    nurse_index_1, _ = get_nurse_and_day(nurse_day_1)
+    # Diagonal term, without the min_duty_days * min_duty_days
+    Q[nurse_day_1, nurse_day_1] += gamma * (1 - (2 * min_duty_days))
+    for nurse_day_2 in range(nurse_day_1 + 1, size):
+        nurse_index_2, _ = get_nurse_and_day(nurse_day_2)
+        if (nurse_index_1 == nurse_index_2 and nurse_day_2 != nurse_day_1):
+            Q[nurse_day_1, nurse_day_2] += gamma * 2
+
+# Solve the problem, and use the offset to scale the energy
+e_offset = (lagrange_parameter * n_days * workforce * workforce) + (gamma * n_nurses * min_duty_days * min_duty_days)
+bqm = BinaryQuadraticModel.from_qubo(Q, offset=e_offset)
+sampler = LeapHybridSampler()
+results = sampler.sample(bqm)
+
+# Get the results
+smpl = results.first.sample
+energy = results.first.energy
+print("Size ", size)
+print("Energy ", energy)
+
+
+# Check the results by doing the sums directly
+# J sum
+sum_j = 0
+for i in range(size):
+    for j in range(size):
+        sum_j += J[i, j] * smpl[i] * smpl[j]
+print("Checking Hard nurse constraint ", sum_j)
+
+sum_w = 0
+# workforce sum
```

Comment:
perhaps move the comment above the `sum_w = 0` line?

Reply:
Thanks.